splitting the field revamp (#3815)
cpinitiative/usaco-guide: solutions/silver/usaco-645.mdx

id: usaco-645
source: USACO Gold 2016 Open
title: Splitting the Field
author: Óscar Garries, Benjamin Qi, Ryan Chou
---

[Official Analysis (Java)](http://www.usaco.org/current/data/sol_split_gold_open16.html)

<Warning title="Weak Test Data">

Some accepted submissions may give different outputs on the following test case:

The right answer should be $6$.

```cpp
// BeginCodeSnip{Test Case}
17 1 1 1 2 1 3 2 1 2 2 2 3 3 1 3 2 3 3 3 4 3 5 4 3 4 4 4 5 5 3 5 4 5 5

>     // EndCodeSnip

Reviewer: sus

Author: noooo

```
</Warning>

## Explanation

> Since the enclosures must be rectangular, there *must* be a vertical or horizontal line that separates the two. Othewise,

Reviewer 2: ```suggestion
Since the enclosures must be rectangular, there is always a vertical or horizontal line that separates the two enclosures. Otherwise,
```

> we'd have a case where one enclosure has to contain some cow on the x-axis *and* some cow on the y-axis as the other enclosure.

Reviewer 2: Wait this doesn't make sense

Reviewer 2: we'd have a case where one enclosure has to contain some cow on the x-axis *and* some cow on the y-axis as the other enclosure.

> Thus, we'll perform two sweeps, checking all possible vertical and horizontal lines. To keep track of the best answer, we'll keep

Reviewer 2: ```suggestion
Thus, we'll perform two sweeps: one checking vertical line and the other checking horizontal lines. To keep track of the best answer, we'll keep
```

prefix and suffix minimums and maximums to calculate each possible answer in $\mathcal{O}(1)$ time.

> It's also possible to use binary search instead of prefix/suffix minimums and maximums.

Reviewer 2: ```suggestion
It's also possible to use binary search instead of prefix and suffix minimums and maximums.
```

Author: don't want consecutive ands, might make it confusing

## Implementation

**Time Complexity:** $\mathcal{O}(N \log N)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

int n;
ll ans = 0;
vector<pair<int, int>> cows;

> // calculate the maximum area saved by testing splits along cows[i].first

Reviewer: make this a doc comment
and i think it should return an int and `main()` should update it instead of how it works currently

	suf[n - 1] = {cows[n - 1].second, cows[n - 1].second};
	for (int i = n - 2; i >= 0; i--) {
		suf[i] = upd(suf[i + 1], cows[i].second);
	}

	// initial area with one enclosure
	ll area = (ll)(cows.back().first - cows.front().first) *
	          (pref.back().second - pref.back().first);

	ll best = LLONG_MAX;
	for (int i = 0; i < n - 1; ++i) {
		// if it's possible to split
		if (cows[i].first != cows[i + 1].first) {
			ll first_rect = (ll)(cows[i].first - cows[0].first) *
			                (pref[i].second - pref[i].first);
			ll second_rect = (ll)(cows.back().first - cows[i + 1].first) *
			                 (suf[i + 1].second - suf[i + 1].first);

			best = min(best, first_rect + second_rect);
		}
	}

	ans = max(ans, area - best);
}

int main() {
	freopen("split.in", "r", stdin);
	freopen("split.out", "w", stdout);

	cin >> n;
	cows.resize(n);

	for (pair<int, int> &cow : cows) { cin >> cow.first >> cow.second; }

	// try all splits on x axis
	search();

	for (pair<int, int> &cow : cows) swap(cow.first, cow.second);

	// try all splits on y axis
	search();

	cout << ans << endl;
}
```
</CPPSection>

<PySection>

```py
import sys

sys.stdin = open("split.in")

> sys.stdout = open("split.out", "w")

Reviewer: can u use `with`?

n = int(input())
cows = [list(map(int, input().split())) for _ in range(n)]
ans = 0

def search():

> 	global ans

Reviewer: doc comment here (use the multistring syntax)

Reviewer: the comment should be below the function def
like look at this: https://stackoverflow.com/questions/2357230/what-is-the-proper-way-to-comment-functions-in-python

Author: omfg im falling off -- thanks <33

	cows.sort()

	upd = lambda x, y: (min(x[0], y), max(x[1], y))

	# compute prefix/suffix minimums and maximums
	pref = [(cows[0][1], cows[0][1])]
	for i in range(1, n):
		pref.append(upd(pref[-1], cows[i][1]))

	suf = [(cows[-1][1], cows[-1][1])]
	for i in range(n - 2, -1, -1):
		suf.append(upd(suf[-1], cows[i][1]))
	suf.reverse()

	# initial area with one enclosure
	area = (cows[-1][0] - cows[0][0]) * (pref[-1][1] - pref[-1][0])
	best = float("inf")

	for i in range(n - 1):
		# if it's possible to split
		if cows[i][0] != cows[i + 1][0]:
			first_rect = (cows[i][0] - cows[0][0]) * (pref[i][1] - pref[i][0])
			second_rect = (cows[-1][0] - cows[i + 1][0]) * (
				suf[i + 1][1] - suf[i + 1][0]
			)

			best = min(best, first_rect + second_rect)

	ans = max(ans, area - best)

# try all splits on x axis
search()
cows = [(y, x) for x, y in cows]
# try all splits on y axis

> search()

Reviewer: space this out?

id: usaco-645
source: USACO Gold 2016 Open
title: Splitting the Field
author: Óscar Garries, Benjamin Qi, Ryan Chou
---

[Official Analysis (Java)](http://www.usaco.org/current/data/sol_split_gold_open16.html)

<Warning title="Weak Test Data">

Some accepted submissions may give different outputs on the following test case:

The right answer should be $6$.

```cpp
// BeginCodeSnip{Test Case}
17 1 1 1 2 1 3 2 1 2 2 2 3 3 1 3 2 3 3 3 4 3 5 4 3 4 4 4 5 5 3 5 4 5 5
    // EndCodeSnip
```
</Warning>

## Explanation

Since the enclosures must be rectangular, there *must* be a vertical or horizontal line that separates the two. Otherwise,
we'd have a case where one enclosure would have to overlap the other in order to enclose all the cows.

![](<usaco-645/enclosures.png>)

Thus, we'll perform two sweeps, one checking vertical splits, and the other checking horizontal splits. To keep track of the best answer, we'll keep
prefix and suffix minimums and maximums to calculate each possible solution in $\mathcal{O}(1)$ time.

> Using binary search instead of prefix/suffix minimums and maximums is also possible.

Reviewer 2: Wait can you explain the binary search part further

Author: sure!

id: usaco-645
source: USACO Gold 2016 Open
title: Splitting the Field
author: Óscar Garries, Benjamin Qi, Ryan Chou
---

[Official Analysis (Java)](http://www.usaco.org/current/data/sol_split_gold_open16.html)

<Warning title="Weak Test Data">

Some accepted submissions may give different outputs on the following test case:

The right answer should be $6$.

> ```

Reviewer: code snip this giant test case

17
1 1
1 2
1 3
2 1
2 2
2 3
3 1
3 2
3 3
3 4
3 5
4 3
4 4
4 5
5 3
5 4
5 5
```
</Warning>

## Explanation

Since the enclosures must be rectangular, there *must* be a vertical or horizontal line that separates the two. Othewise,
we'd have a case where one enclosure has to contain some cow on the x-axis *and* some cow on the y-axis as the other enclosure.

Thus, we'll perform two sweeps, checking all possible vertical and horizontal lines. To keep track of the best answer, we'll keep
prefix and suffix minimums and maximums to calculate each possible answer in $\mathcal{O}(1)$ time.

It's also possible to use binary search instead of prefix/suffix minimums and maximums.

## Implementation

**Time Complexity:** $\mathcal{O}(N \log N)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

> int N;

Reviewer: rename to n

ll ans = 0;
vector<pair<int, int>> cows;

> void search() {

Reviewer: doc comment

	sort(cows.begin(), cows.end());

	vector<pair<int, int>> pref(N), suf(N);

	auto upd = [](pair<int, int> a, int b) -> pair<int, int> {
		return {min(a.first, b), max(a.second, b)};

> 	};

Reviewer: what this do

	// initial area with one enclosure
	ll area = (ll)(cows.back().first - cows.front().first) *
	          (pref.back().second - pref.back().first);

	ll best = LLONG_MAX;

> 	for (int i = 0; i < n - 1; ++i) {

Reviewer: make this i++ instead of ++i

		suf[i] = upd(suf[i + 1], cows[i].second);
	}

	// initial area with one enclosure
	ll area = (ll)(cows.back().first - cows.front().first) *
	          (pref.back().second - pref.back().first);

	ll best = LLONG_MAX;
	for (int i = 0; i < N - 1; ++i) {
		// if it's possible to split
		if (cows[i].first != cows[i + 1].first) {
			ll first_rect = (ll)(cows[i].first - cows[0].first) *
			                (pref[i].second - pref[i].first);
			ll second_rect = (ll)(cows.back().first - cows[i + 1].first) *
			                 (suf[i + 1].second - suf[i + 1].first);

			best = min(best, first_rect + second_rect);
		}
	}

	ans = max(ans, area - best);
}

int main() {
	freopen("split.in", "r", stdin);
	freopen("split.out", "w", stdout);

	cin >> N;
	cows.resize(N);

	for (pair<int, int> &cow : cows) cin >> cow.first >> cow.second;

	// try all splits on x axis
	search();

> 	for (pair<int, int> &cow : cows) swap(cow.first, cow.second);

Reviewer: braces

Reviewer: not fixed?

Author: im so sorry i thought i did it :<<
thanks

<PySection>

```py
import sys

sys.stdin = open("split.in")

> sys.stdout = open("split.out", "w")

Reviewer: is this still needed
